lib/cli/run.js

```diff
@@ -368,50 +368,22 @@ exports.builder = yargs =>
       validatePlugin(argv, 'reporter', Mocha.reporters);
       validatePlugin(argv, 'ui', Mocha.interfaces);
 
+      // post-process options
+      if (argv.noTimeouts) {
+        argv.enableTimeouts = !argv.noTimeouts;
+      }
+
+      if (argv.retries === 0) {
+        delete argv.retries;
+      }
+
       return true;
     });
 
 exports.handler = argv => {
   debug('post-yargs config', argv);
 
-  const mocha = new Mocha();
-
-  mocha
-    .bail(argv.bail)
-    .fgrep(argv.fgrep)
-    .globals(argv.globals)
-    .grep(argv.grep)
-    .reporter(argv.reporter, argv.reporterOptions)
-    .slow(argv.slow)
-    .timeout(argv.timeout)
-    .ui(argv.ui)
-    .useColors(argv.colors)
-    .useInlineDiffs(argv.inlineDiffs);
-
-  // pesky inverted options
-  mocha.hideDiff(!argv.diff).enableTimeouts(!argv.noTimeouts);
-
-  // pain: toggles NOT accepting parameters
-  [
-    'allowUncaught',
-    'asyncOnly',
-    'checkLeaks',
-    'delay',
-    'forbidOnly',
-    'forbidPending',
-    'fullTrace',
-    'growl',
-    'invert'
-  ].forEach(arg => {
-    if (argv[arg]) {
-      mocha[arg]();
-    }
-  });
-
-  // more pain: retries() is nonsensical
-  if ('retries' in argv) {
-    mocha.retries(argv.retries);
-  }
+  const mocha = new Mocha(argv);
```

Comment:
Shouldn't you split the `argv` object up into relevant parts before use? And why would `runMocha` need `argv`?
<https://github.com/mochajs/mocha/blob/3f046723e7171c0646a065315cbef078fb03617f/lib/cli/run.js#L383-L392>

Reply:
Well, what the Mocha ctor takes as an argument is a plain object containing properties. Not what you gave it. Same with `handleFiles`, which has no need of `argv` save its positional arguments.

_actually_

Reply:
FWIW `handleFiles` needs quite a bit more than the positional args:

```js
/**
 * Smash together an array of test files in the correct order
 * @param {Object} [opts] - Options
 * @param {string[]} [opts.extension] - File extensions to use
 * @param {string[]} [opts.spec] - Files, dirs, globs to run
 * @param {string[]} [opts.exclude] - Files, dirs, globs to exclude
 * @param {boolean} [opts.recursive=false] - Find files recursively
 * @param {boolean} [opts.sort=false] - Sort test files
 * @returns {string[]} List of files to test
 * @private
 */
```

> I did this same transition a year ago. I learned it becomes a PITA to simulate stuff generating fake yargs structures for testing instead of just pruning down and passing functions what they really required (and no more).

I still don't quite understand.  

They don't actually have to *be* `yargs.Arguments` objects, that's true.  There's nothing really special about this object other than it has a convention for positional arguments--which we're not consuming anywhere (see `spec` in the docstring above).

What it *sounds* like you're suggesting is extracting all of the properties that are relevant to each function and passing *only those* instead of the whole thing.  If, for some reason, we needed a new option passed to one of these functions, we'd have it already.  If we didn't, as suggested, it'd create churn in the code that filters properties...

What problems were you encountering in your tests?